lib/server-app.js

```diff
@@ -161,7 +161,28 @@ proto.middleware = function(name, paths, handler) {
 
   this._skipLayerSorting = true;
   this.use(paths, handler);
-  this._router.stack[this._router.stack.length - 1].phase = fullPhaseName;
+
+  // Other handlers can be added to the stack, for example,
+  // NewRelic adds sentinel handler. We need to search the stack
+  var found = false;
+  for (var k = this._router.stack.length - 1; k >= 0 && !found; k--) {
+    if (this._router.stack[k].handle === handler ||
+      // NewRelic replaces the handle and keeps it as __NR_original
+      this._router.stack[k].handle['__NR_original'] === handler
+      ) {
+      this._router.stack[k].phase = fullPhaseName;
+      found = true;
+    } else {
+      // Agressively check if the original handler has been wrapped
+      // into a new function with a property pointing to the original handler
+      for (var p in this._router.stack[k].handle) {
+        if (this._router.stack[k].handle[p] === handler) {
+          this._router.stack[k].phase = fullPhaseName;
+          found = true;
+        }
+      }
+    }
+  }
```

Comment:
Please extract this new block into a standalone function. Perhaps a function that accepts the `handler` and returns the appropriate stack layer:

```
// usage
var layer = findHandlerLayer(this._router.stack, handler);
if (layer) {
  layer.phase = fullPhaseName;
} else {
  /* print a warning? */ 
}
```

Add unit tests for the two new scenarios:

 - the handler function is stored in `handle[__NR_original]`
 - the handler function is stored in a different property